pkg/cmd/roachtest/tests/mvcc_gc.go

```diff
@@ -0,0 +1,334 @@
+		databaseName, tableName, strings.Join(values, ", "), int(expiration.Seconds()))
+	if _, err := conn.Exec(stmt); err != nil {
+		t.Fatal("failed to split kv table into ranges", err)
+	}
+}
+
+func assertTableHasNoRangeTombstones(
+	t test.Test, conn *gosql.DB, tableName, databaseName string,
+) error {
+	totals := collectTableMVCCStats(t, conn, tableName, databaseName)
+	t.L().Printf("Assert range tombstones in stats: %s", totals)
+	if totals["RangeKeyCount"] > 0 || totals["RangeKeyBytes"] > 0 || totals["Ranges"] == 0 {
+		return errors.Errorf("table ranges contain range tombstones %s", totals.String())
+	}
+	return nil
+}
+
+func assertTableHasNoLiveData(t test.Test, conn *gosql.DB, tableName, databaseName string) error {
+	totals := collectTableMVCCStats(t, conn, tableName, databaseName)
+	t.L().Printf("Assert live data in stats: %s", totals)
+	if totals["LiveBytes"] > 0 || totals["LiveCount"] > 0 ||
+		totals["IntentBytes"] > 0 || totals["IntentCount"] > 0 || totals["SeparatedIntentCount"] > 0 ||
+		totals["Ranges"] == 0 {
+		return errors.Errorf("table ranges contain live data %s", totals.String())
+	}
+	return nil
+}
+
+func collectTableMVCCStats(
+	t test.Test, conn *gosql.DB, tableName, databaseName string,
+) testRangeStats {
+	t.L().Printf("Rescanning table ranges for garbage")
+	row := conn.QueryRow(`select min(start_key), max(end_key) from crdb_internal.ranges_no_leases where database_name = $1 and table_name = $2`,
+		tableName, databaseName)
+	if row.Err() != nil {
+		t.Fatal("failed to query key boundaries for table", row.Err())
+	}
+	var startKey, endKey roachpb.Key
+	if err := row.Scan(&startKey, &endKey); err != nil {
+		t.Fatal("failed to query key boundaries for table", row)
+	}
+	rows, err := conn.Query(`select range_id, status, detail from crdb_internal.check_consistency(true, $1, $2) order by start_key`, startKey, endKey)
+	if err != nil {
+		t.Fatal("failed to run consistency check query on table", err)
+	}
+	defer rows.Close()
+
+	var result = make(testRangeStats)
+	for rows.Next() {
+		var (
+			rangeID int
+			status  string
+			detail  string
+		)
+		if err := rows.Scan(&rangeID, &status, &detail); err != nil {
+			t.Fatal("failed to run consistency check query on table", err)
+		}
+		stats := getRangeStatsFromDetail(detail)
+		result.merge(stats)
+	}
+	return result
+}
+
+type testRangeStats map[string]int
+
+func (s testRangeStats) String() string {
+	keys := make([]string, len(s))
+	i := 0
+	for k := range s {
+		keys[i] = k
+		i++
+	}
+	sort.Strings(keys)
+	stats := make([]string, len(keys))
+	for i, k := range keys {
+		stats[i] = fmt.Sprintf("%s: %d", k, s[k])
+	}
+	return fmt.Sprintf("{%s}", strings.Join(stats, " "))
+}
+
+func (s testRangeStats) merge(other testRangeStats) {
+	for k, v := range other {
+		s[k] += v
+	}
+}
+
+var statsRE = regexp.MustCompile(`.*stats: {([^}]*)}.*`)
+var valuesRE = regexp.MustCompile(`(\w+):(\d+)`)
+
+func getRangeStatsFromDetail(detail string) testRangeStats {
```

Comment:
This seems really brittle. I don't know if we have a proper API to retrieve MVCC stats via SQL. If not, we could either add one, or use the status server's `Range` endpoint.